Naming article, part 1

tjstankus opened the pull request.

tjstankus committed:
Add abstraction ladder intro and temp image

tjstankus committed:
Add prose for abstraction ladder

tjstankus committed:
Edit and format

I'll unwrap paragraphs when it's ready to publish.

tjstankus committed:
Make progress on draft

tjstankus committed:
Something like a draft

tjstankus committed:
Make edits from printed readthrough

The idea of two rules and other changes are up next.

tjstankus committed:
People like rules

tjstankus committed:
Expand guidelines to 4

Needs feedback

tjstankus committed:
Address todos, make diagrams

tjstankus committed:
Update images, spell check, add guidelines summary

tjstankus committed:
Spell check

tjstankus committed:
Update pandoc instructions

tjstankus committed:
Self copy edit

tjstankus committed:
Restructure per feedback

Symmetry: each "rung" section ends with a guideline. Removed the
"context matters" guideline. Added a guideline re: describing
implementation details. Some rewording, esp. around transitions.

tjstankus committed:
Update pandoc instructions

tjstankus committed:
Move everything into article-specific directory

tjstankus commented on 2019-12-naming-part1/naming.md:
```diff
@@ -0,0 +1,231 @@
+
+If you can trust that a library works, you need look no further than its API docs to understand it. Ideally, API docs are the only point of contact you have with a third-party library. And they are, essentially, a bunch of names.
+
+Just as a writer grinds through word choices, a programmer must struggle with names. As the joke goes, the two hardest things in computer science are naming, cache invalidation, and off-by-one errors. No matter the programming language, no matter whether you're operating from a functional or object-oriented perspective, no matter if the language is statically or dynamically typed, names are a ubi ...
```
Here's my feedback for this paragraph.

tjstankus commented on 2019-12-naming-part1/naming.md:
```diff
@@ -0,0 +1,231 @@
+
+### Bottom rung: dirty mirrors
+
+You have an implementation that works. You have, of course, written tests (not shown). In your application, there's a `PersonPresenter` module, in which this functionality belongs. But, now you must create a function and name it.
```
Test comment.

tjstankus commented:
General comment.

tjstankus commented on 2019-12-naming-part1/naming.md:
```diff
@@ -0,0 +1,231 @@
+
+But naming demands that you wrest your mind from granular, concrete details and think more generally, more categorically.
+
+The linguist Sam Hayakawa, in his book [Language in Thought and Action](https://www.amazon.com/Language-Thought-Action-S-I-Hayakawa/dp/0156482401/), describes a model of language he calls the Abstraction Ladder. On this conceptual ladder, the bottommost rung holds the most concrete idea. The topmost rung is occupied by a decidedly abstract, intangible concept. In between, ideas get more abstract  ...
```
A comment from me

tjstankus commented:
Testing

trimm002 commented:
Checking out this "general comment" feature. Looks like I've got it!

trimm002 commented on 2019-12-naming-part1/naming.md:
```diff
@@ -0,0 +1,231 @@
+
+The linguist Sam Hayakawa, in his book [Language in Thought and Action](https://www.amazon.com/Language-Thought-Action-S-I-Hayakawa/dp/0156482401/), describes a model of language he calls the Abstraction Ladder. On this conceptual ladder, the bottommost rung holds the most concrete idea. The topmost rung is occupied by a decidedly abstract, intangible concept. In between, ideas get more abstract  ...
+
+<img src="images/abstraction-ladder.svg" width="100%" alt="The Abstraction Ladder" />
```
This is a great diagram but I can't figure out what the bottom-most image is: a pair of lungs? Mittens and a zipper? Just wondering.

tjstankus replied:
Ha, now I'm wondering! I think it's lungs and a spine. At least that's the story I'm telling! I clarified in the caption in this commit: 1ec077264e27e2c778857aa09c3c3ac32d128b08.

tjstankus replied:
What I *thought* might get someone's pushback is the idea of a rat as a pet. But I'm going to double down on that!

trimm002 commented on 2019-12-naming-part1/naming.md:
```diff
@@ -0,0 +1,231 @@
+
+<img src="images/abstraction-ladder.svg" width="100%" alt="The Abstraction Ladder" />
+
+Consider the diagram above. Molly is on the second rung. She is far from a static entity, she's a being in-process, constantly changing, composed of numerous subsystems that are represented on the bottom rung. But from a cognitive, linguistic perspective, Molly seems like a single entity. Indeed, associating a name with a living being is often the first utterance of a toddler learning to speak -  ...
```
Two suggestions: separate the clauses with an em-dash and remove the hypen from in-process (it's used as a noun phrase here so no need for the hypen). "She is far from a static entity -- she's a being in process"

tjstankus replied:
Addressed in e2195e047595a91cdbc86b740c6e85f0250d5d13

trimm002 commented on 2019-12-naming-part1/naming.md:
```diff
@@ -0,0 +1,231 @@
+
+Consider the diagram above. Molly is on the second rung. She is far from a static entity, she's a being in-process, constantly changing, composed of numerous subsystems that are represented on the bottom rung. But from a cognitive, linguistic perspective, Molly seems like a single entity. Indeed, associating a name with a living being is often the first utterance of a toddler learning to speak -  ...
+
+The next rung up the ladder, "Dogs" seems like a small step. But, consider the enormous mental processing it takes for a toddler to learn to ignore the differences between the family dog Molly, a small bull terrier, and the gigantic great dane that lives down the street. Learning to abstract from "Molly" to "Dogs" takes significant mental development, but once it's learned, it's automatic, proces ...
```
Think it should be "Great Dane."

tjstankus replied:
Addressed in e2195e047595a91cdbc86b740c6e85f0250d5d13

trimm002 commented on 2019-12-naming-part1/naming.md:
```diff
@@ -0,0 +1,231 @@
+
+## Naming functions
+
+Communication is most effective when it spans various levels of abstraction, when it travels up and down the ladder. Just stating that you should consider abstract ideas when naming things in your code is not enough. Providing abstract naming advice also requires concrete examples in order to be convincing. Concrete examples explain abstract ideas. In order to explain the color green, you might u ...
```
Instead of "Just stating . . . " how about this: "It's not enough to recommend that you consider abstract ideas when naming things in your code." (suggestion for removing the word "Just" which most of us overuse.)

tjstankus replied:
Addressed in e2195e047595a91cdbc86b740c6e85f0250d5d13

tjstankus committed:
Integrate Julia's recommended text changes

tjstankus committed:
Clarify bottom image on abstraction ladder

trimm002 commented on 2019-12-naming-part1/naming.md:
````diff
@@ -0,0 +1,231 @@
+|> String.split()
+|> Enum.join(" ")
+```
+
````
I might suggest a short transition sentence to the next section that invites the reader to continue with the hypothetical implementation. Something like "Your hypothetical application needs to handle this unfortunate data gracefully."

tjstankus replied:
Changed in 745892680cded5783f9c6b545ed14fec6949b189

trimm002 commented on 2019-12-naming-part1/naming.md:
````diff
@@ -0,0 +1,231 @@
+end
+```
+
+At this point, it might be tempting to reach for an overtly literal function name, like `split_and_join` above. Indeed, the code splits and joins a name. It's quick and easy to conceive of; it's convenient.
````
This was the only spot that distracted me. I think the semicolon is too jarring for this sentence It's quick and easy to conceive of; it's convenient. I might prefer "It's quick and easy to conceive of -- it's a convenient option." Might be good to emphasize that it's a little too easy to devise. It's the lazy choice.

tjstankus replied:
Addressed in bb7a6a3a9ed344ecafaedf8ebe2ea5d183cf6f35

trimm002 commented on 2019-12-naming-part1/naming.md:
````diff
@@ -0,0 +1,231 @@
+end
+```
+
+This is better. But recall that the data comes from a source you do not own. It's controlled by someone else and likely to change in unpredictable ways. Even code and data that you *do* own are susceptible to ever-evolving business requirements. The code you write today might seem concrete, but like Molly the dog, even seemingly concrete entities are not static. Code is a being in-process, consta ...
````
It's really OK to start a sentence with "But" occasionally. But you've done it 13 times in this blog post. I might suggest that it's a habit of yours. Maybe aim for three usages, in places where you want a particularly hard stop. Try combining with the previous sentence, "This is better, but recall that the data comes from a source you do not own." Resist the urge to substitute But with However or Actually since they are overused in general.

tjstankus replied:
😄 I literally laughed out loud at the second sentence. Wow. Oh my, even. Great catch. I didn't realize I was doing that. I'll go through and implement your suggestions.

tjstankus replied:
Addressed in 075a1ac8daa99338b6f2d9cf8a840998173f1678

trimm002 commented on 2019-12-naming-part1/naming.md:
```diff
@@ -0,0 +1,231 @@
+
+Should you rename the function, perhaps to `trim_whitespace_and_reorder`? Down that path lies madness. Naming a function after its implementation leads to expensive, hand-wringing decisions.
+
+But this simple example obscures an even more costly consideration. Your codebase depends on many messaging interfaces, as it should. In order to get anything done, your code must send messages somewhere. In other words, it must know the names of functions it can call. When your code calls a function, it becomes dependent upon the name of that function. Again, this is fine. It's better than fine; ...
```
You could say, "And this simple example obscures . . . "

tjstankus replied:
Resolved in 345a1339f8f2b4e506e3897474916427967dd070

trimm002 commented on 2019-12-naming-part1/naming.md:
```diff
@@ -0,0 +1,231 @@
+
+But this simple example obscures an even more costly consideration. Your codebase depends on many messaging interfaces, as it should. In order to get anything done, your code must send messages somewhere. In other words, it must know the names of functions it can call. When your code calls a function, it becomes dependent upon the name of that function. Again, this is fine. It's better than fine; ...
+
+In your hypothetical app, a person's name is presented in several views. `trim_whitespace` is called from many places. Your codebase has become dependent upon that name. Alarm bells should be going off in your head when you see names representing implementation details spread throughout your codebase. Changing this function name at its source means changing it everywhere it's called. There's a co ...
```
Where does "Change Preventer" come from? Not another code smell? Your use of capitals suggests that it's a proper noun.

tjstankus replied:
It's probably a good idea to link to information on Shotgun Surgery and Change Preventer. (Note to self: refactoring.com does not have code smell categories. The categories stem from a research paper that Sandi mentioned in one of her talks, so perhaps just link to that talk. Introduce the idea of categories more explicitly.)

tjstankus replied:
Addressed in ac3fc886375a96bd557de03835275634397a4d63

trimm002 commented on 2019-12-naming-part1/naming.md:
```diff
@@ -0,0 +1,231 @@
+
+You might decide to live with the lie, implementing new functionality without changing the function name. Your name then professes that the function behaves in a way that it actually doesn't. This decision is the quickest to implement. It might appease your product team, but in terms of software engineering, it's kicking the can down the road. You're accruing technical debt and confusing your pee ...
+
+The cost of code is in the reading. Your code will be expensive to read as long as it is poorly named. But living with the lie and conceding technical debt, in this case, is the optimist's view. Changing a function's behavior to betray its name almost inevitably leads to downstream bugs. It's not hard to imagine a future programmer, maybe even you, deciding to trust a name and being burned with b ...
```
I'd just remove the "But" in this third sentence.

tjstankus replied:
Resolved in 345a1339f8f2b4e506e3897474916427967dd070

trimm002 commented on 2019-12-naming-part1/naming.md:
```diff
@@ -0,0 +1,231 @@
+
+You might determine that the cost of implementing the new requirement is too high in the present and poses too much risk in the near future. You'd prefer to prevent the change altogether. When you push back on the product team, explaining that it's too complex and expensive to make now, expect looks of befuddlement in response. Those looks are justified. A simple request should be simple to imple ...
+
+This might seem like hyperbole. How can a single poorly named function make your programming life unpleasant and engender organizational mistrust in software engineering? A single poorly named function probably won't. But, the *habit* of naming functions after their current implementation will infect your codebase, making all aspects of development slower, more expensive, and less fun.
```
This is a fine sentence, well-constructed and easy to read. "A single poorly named function probably won't, but the *habit* of naming functions after their current implementation will infect your codebase, making all aspects of development slower, more expensive, and less fun.

trimm002 commented on 2019-12-naming-part1/naming.md:
```diff
@@ -0,0 +1,231 @@
+
+Function names should tell the right story. Good names are resilient &mdash; they stand the test of time through changes in implementation. Naming is difficult, but names matter. It's incumbent upon you to develop this core programming skill. So, how should you name functions? 
+
+`trim_whitespace` is clearly named after its current implementation. The requirement to reorder the name to `last, first` order has forced you into a decision: to rename or not to rename. But, that decision could have been avoided altogether, by wresting your mind from granular implementation details to think about names more categorically. Your function name should move one rung higher.
```
You could replace "But," with "However," in this paragraph.

tjstankus replied:
Resolved in 345a1339f8f2b4e506e3897474916427967dd070

trimm002 commented on 2019-12-naming-part1/naming.md:
```diff
@@ -0,0 +1,231 @@
+| Capitalize words             | `capitalize`      |
+| Prepend salutation           | `add_salutation`  |
+
+The verb `present` is a viable candidate for the missing column heading in the matrix above. But the name of the module is `PersonPresenter`. Using the word `present` would read like an echo chamber: `PersonPresenter.present_name`.
```
Remember that the name of the module is `PersonPresenter` so naming the function `present` would read like an echo chamber: `PersonPresenter.present_name`.

tjstankus replied:
Resolved in 345a1339f8f2b4e506e3897474916427967dd070

trimm002 commented on 2019-12-naming-part1/naming.md:
```diff
@@ -0,0 +1,231 @@
+
+**GUIDELINE: Name functions one level of abstraction higher than their implementation**
+
+### Climb higher: ask a friend for stability
```
I think this ending is strong. It flows wel.

trimm002 replied:
oop! "well"

trimm002 commented on 2019-12-naming-part1/naming.md:
```diff
@@ -0,0 +1,231 @@
+
+## Climb up and down the ladder
+
+Does software design really demand this level of attention to detail, this careful weighing of names for something as unremarkable and ordinary as a single function? If you want to do your best work, yes, it does. Names matter. Attention to details matters. Writing software is about more than just getting something to work. At an abstract level, it's an expression of thought and communication, li ...
```
Maybe needs "then":  "If you want to do your best work, then yes, it does."

tjstankus replied:
Resolved in 345a1339f8f2b4e506e3897474916427967dd070

trimm002 commented on 2019-12-naming-part1/naming.md:
```diff
@@ -0,0 +1,231 @@
+
+A literal function name indicates that you know what a function does but may not have considered what it means or how it's perceived by others who use it. When your internal and external dialog spans multiple levels of the Abstraction Ladder, you become open to the full breadth of understanding that categorical thinking bestows. That deeper understanding leads to better code, in every way.
+
+You have guidelines for improving names and your code will be better for following them. But how do you help others improve? How do you spread value throughout your organization? Part 2 of this article will explore those questions and offer advice about naming classes and modules.
```
"Thinking more broadly, how do you help others improve?"

tjstankus replied:
Resolved in 345a1339f8f2b4e506e3897474916427967dd070

trimm002 commented on 2019-12-naming-part1/naming.md:
```diff
@@ -0,0 +1,231 @@
+
+You have guidelines for improving names and your code will be better for following them. But how do you help others improve? How do you spread value throughout your organization? Part 2 of this article will explore those questions and offer advice about naming classes and modules.
+
+The ideas and techniques in this article rely and expand upon the content in the [99 Bottles of OOP book](https://99bottlesofoop.com) and [Sandi Metz's Practical Object-Oriented Design Course](https://www.sandimetz.com/courses). The second edition of the 99 Bottles book is in the works! It will contain several new chapters with about 50% more content than the first edition. Now, when you buy the  ...
```
I'd like to see that exclamation point downgraded to a period. It's exciting but it's not exclamatory. Not a big deal though.

tjstankus replied:
Resoved in 345a1339f8f2b4e506e3897474916427967dd070

tjstankus committed:
Make easy changes based on feedback

I'll address the ones that will take more time next.

tjstankus commented:
@trimm002 I went through and integrated the changes I could do quickly. When I have a longer time block, I'll tackle the rest. Thanks for the great suggestions!

tjstankus commented:
More notes to self:
- At Coupa, one idea that the students latched onto was, "If you can't come up with good names, you don't understand your code."
- These guidelines are by no means an exhaustive list. There are infinite considerations and unique circumstances. There is no one-size-fits-all list of advice.

tjstankus committed:
Remove a 'you', add transition sentence

tjstankus committed:
Remove semicolon, add "lazy"

One of these days I will deploy a semicolon with laser precision and to
universal acclaim.

tjstankus committed:
Link, attribute references

tjstankus committed:
Add to wrap-up based on Coupa experiences

The students at Coupa really latched on to "If you can't come up with
good names, you don't understand your code."

tjstankus committed:
Reduce overuse of But

I like capital-B Buts and I cannot lie

trimm002 commented on 2019-12-naming-part1/naming.md:
```diff
@@ -0,0 +1,234 @@
+
+Should you need to change how the name is formatted in response to upstream data changes or new application requirements, the impact of the implementation will be constrained to the internals of the function. The name will continue to be appropriate and no dependents will be forced to change.
+
+In terms of the Open Closed Principle, the code is more open to the change. For example, a new requirement to add a salutation would only necessitate changes to implementation internals, no API-breaking name change would be necessary. By increasing the level of abstraction, you have created an approprate and resilient function name.
```
"appropriate" is misspelled in the last sentence (as "approprate")

tjstankus replied:
Good catch! Resolved in 953c87eedef5e23f55300d93cb56f52f381f0f6c.

trimm002 commented on 2019-12-naming-part1/naming.md:
```diff
@@ -0,0 +1,234 @@
+
+<img src="images/abstraction-ladder.svg" width="100%" alt="The Abstraction Ladder" />
+
+Consider the diagram above. Molly is on the second rung. She is far from a static entity &mdash; she's a being in process, constantly changing, composed of numerous subsystems that are represented on the bottom rung. From a cognitive, linguistic perspective, however, Molly seems like a single entity. Indeed, associating a name with a living being is often the first utterance of a toddler learning ...
```
This is very minor but I'd put an em dash here "toddler learning to speak -- "Mama."

tjstankus replied:
Resolved in 953c87eedef5e23f55300d93cb56f52f381f0f6c.

trimm002 commented on 2019-12-naming-part1/naming.md:
```diff
@@ -0,0 +1,234 @@
+
+You might determine that the cost of implementing the new requirement is too high in the present and poses too much risk in the near future. You'd prefer to prevent the change altogether. When you push back on the product team, explaining that it's too complex and expensive to make now, expect looks of befuddlement in response. Those looks are justified. A simple request should be simple to imple ...
+
+This might seem like hyperbole. How can a single poorly named function make your programming life unpleasant and engender organizational mistrust in software engineering? A single poorly named function probably won't. But, the *habit* of naming functions after their current implementation will infect your codebase, making all aspects of development slower, more expensive, and less fun.
```
No need for a comma after the "But" here.

tjstankus replied:
Resolved in 953c87eedef5e23f55300d93cb56f52f381f0f6c.

trimm002 commented on 2019-12-naming-part1/naming.md:
```diff
@@ -0,0 +1,234 @@
+
+Context and perspective also dictate that the function be named not just `format`, but `format_name` Why include `_name` when the function will accept `name` as a parameter? Despite the fact that it, too, creates a bit of an echo chamber, it contains the right amount of information from the caller's perspective. Right away you know what this does: `PersonPresenter.format_name("  Jane  Doe ")`. Yo ...
+
+If the module were named `PersonFormatter`, it might make more sense to name the function something else, perhaps simply `name`. Naming the module is outside the purview of this article, but will be addressed in a followup article.
```
I might say "follow-up" instead of "followup."

trimm002 commented on 2019-12-naming-part1/naming.md:
```diff
@@ -0,0 +1,234 @@
+
+A literal function name indicates that you know what a function does but may not have considered what it means or how it's perceived by others who use it. Poor names betray a lack of understanding. When your internal and external dialog spans multiple levels of the Abstraction Ladder, you become open to the full breadth of understanding that categorical thinking bestows. That deeper understanding ...
+
+The guidelines in this article are by no means an exhaustive list. Naming presents infinite considerations and unique circumstances. There is no one-size-fits-all list of advice. Yet, you have some guidelines for improving names and your code will be better for considering them. Thinking more broadly, how do you help others improve? How do you spread value throughout your organization? Part 2 of  ...
```
How about saying, "Part 2 of this article, coming in the next blog post, will explore . . ."

tjstankus replied:
I'm not sure that will be the next blog post. :)

tjstankus committed:
Make changes suggested in final copy edit

tjstankus committed:
Add mailing draft

tjstankus committed:
Add 'continuous learning'

tjstankus committed:
Add missing space

skmetz committed:
tweaks to mailing text

skmetz committed:
Merge pull request #2 from codedevotional/naming-mailing-tweaks

tweaks to mailing text

skmetz committed:
Update mailing.md
